core/src/main/java/org/apache/iceberg/deletes/Deletes.java

```diff
@@ -137,93 +150,91 @@ protected boolean shouldKeep(T row) {
     private final CloseableIterable<T> rows;
-    private final Function<T, Long> extractPos;
-    private final CloseableIterable<Long> deletePositions;
+    private final CloseableIterator<Long> deletePosIterator;
+    private final Function<T, Long> rowToPosition;
+    private long nextDeletePos;
 
-    private PositionStreamDeleteFilter(CloseableIterable<T> rows, Function<T, Long> extractPos,
-                                       CloseableIterable<Long> deletePositions) {
+    PositionStreamDeleteIterable(CloseableIterable<T> rows, Function<T, Long> rowToPosition,
+                                 CloseableIterable<Long> deletePositions) {
       this.rows = rows;
-      this.extractPos = extractPos;
-      this.deletePositions = deletePositions;
+      this.rowToPosition = rowToPosition;
+      this.deletePosIterator = deletePositions.iterator();
     }
 
     @Override
     public CloseableIterator<T> iterator() {
-      CloseableIterator<Long> deletePosIterator = deletePositions.iterator();
-
       CloseableIterator<T> iter;
       if (deletePosIterator.hasNext()) {
-        iter = new PositionFilterIterator(rows.iterator(), deletePosIterator);
+        nextDeletePos = deletePosIterator.next();
+        iter = createPosDeleteIterator(rows.iterator());
       } else {
         iter = rows.iterator();
-        try {
-          deletePosIterator.close();
-        } catch (IOException e) {
-          throw new UncheckedIOException("Failed to close delete positions iterator", e);
-        }
       }
 
       addCloseable(iter);
+      addCloseable(deletePosIterator);
 
       return iter;
     }
 
-    private class PositionFilterIterator extends FilterIterator<T> {
-      private final CloseableIterator<Long> deletePosIterator;
-      private long nextDeletePos;
+    boolean isDeleted(T row) {
+      long currentPos = rowToPosition.apply(row);
+      if (currentPos < nextDeletePos) {
+        return false;
+      }
 
-      protected PositionFilterIterator(CloseableIterator<T> items, CloseableIterator<Long> deletePositions) {
-        super(items);
-        this.deletePosIterator = deletePositions;
+      // consume delete positions until the next is past the current position
+      boolean isDeleted = currentPos == nextDeletePos;
+      while (deletePosIterator.hasNext() && nextDeletePos <= currentPos) {
         this.nextDeletePos = deletePosIterator.next();
+        if (!isDeleted && currentPos == nextDeletePos) {
+          // if any delete position matches the current position
+          isDeleted = true;
+        }
       }
 
-      @Override
-      protected boolean shouldKeep(T row) {
-        long currentPos = extractPos.apply(row);
-        if (currentPos < nextDeletePos) {
-          return true;
-        }
+      return isDeleted;
+    }
+
+    protected abstract CloseableIterator createPosDeleteIterator(CloseableIterator<T> items);
```

Comment:
Do we really create a position delete iterator? Don't we iterate over records here?

Reply:
We do iterate the pos delete records, for example, line 191, `this.nextDeletePos = deletePosIterator.next();`.

Reply:
We do iterate over position deletes but I think this iterate is for remaining data records, no?

Reply:
Discussed offline with @aokolnychyi, I've changed the name to `applyDelete`